src/main/java/net/dries007/tfc/common/blocks/PitKilnBlock.java

```diff
@@ -0,0 +1,207 @@
+package net.dries007.tfc.common.blocks;
+
+import java.util.Random;
+
+import net.minecraft.block.Block;
+import net.minecraft.block.BlockState;
+import net.minecraft.block.Blocks;
+import net.minecraft.entity.player.PlayerEntity;
+import net.minecraft.item.ItemStack;
+import net.minecraft.particles.ParticleTypes;
+import net.minecraft.state.IntegerProperty;
+import net.minecraft.state.StateContainer;
+import net.minecraft.util.*;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.BlockRayTraceResult;
+import net.minecraft.util.math.shapes.ISelectionContext;
+import net.minecraft.util.math.shapes.VoxelShape;
+import net.minecraft.util.math.shapes.VoxelShapes;
+import net.minecraft.world.IBlockReader;
+import net.minecraft.world.IWorld;
+import net.minecraft.world.IWorldReader;
+import net.minecraft.world.World;
+
+import net.minecraftforge.items.ItemHandlerHelper;
+
+import net.dries007.tfc.common.TFCTags;
+import net.dries007.tfc.common.tileentity.InventoryTileEntity;
+import net.dries007.tfc.common.tileentity.PitKilnTileEntity;
+import net.dries007.tfc.util.Helpers;
+
+public class PitKilnBlock extends Block implements IForgeBlockProperties
+{
+    private final ForgeBlockProperties properties;
+
+    public static final IntegerProperty STAGE = TFCBlockStateProperties.PIT_KILN_STAGE;
+
+    private static final VoxelShape ONE_SIXTEENTH = Block.box(0.0D, 0.0D, 0.0D, 16.0D, 1.0D, 16.0D);
+    private static final VoxelShape ONE_EIGHTH = Block.box(0.0D, 0.0D, 0.0D, 16.0D, 2.0D, 16.0D);
+    private static final VoxelShape THREE_SIXTEENTHS = Block.box(0.0D, 0.0D, 0.0D, 16.0D, 3.0D, 16.0D);
+    private static final VoxelShape ONE_FOURTH = Block.box(0.0D, 0.0D, 0.0D, 16.0D, 4.0D, 16.0D);
+    private static final VoxelShape FIVE_SIXTEENTHS = Block.box(0.0D, 0.0D, 0.0D, 16.0D, 5.0D, 16.0D);
+    private static final VoxelShape THREE_EIGHTHS = Block.box(0.0D, 0.0D, 0.0D, 16.0D, 6.0D, 16.0D);
+    private static final VoxelShape SEVEN_SIXTEENTHS = Block.box(0.0D, 0.0D, 0.0D, 16.0D, 7.0D, 16.0D);
+    private static final VoxelShape ONE_HALF = Block.box(0.0D, 0.0D, 0.0D, 16.0D, 8.0D, 16.0D);
+    private static final VoxelShape THREE_FOURTHS = Block.box(0.0D, 0.0D, 0.0D, 16.0D, 12.0D, 16.0D);
+
+    public static final int STRAW_END = 7;
+    public static final int LOG_START = 8;
+    public static final int LIT = 16;
+
+    protected static final VoxelShape[] SHAPE_BY_LAYER = new VoxelShape[] {
+        ONE_SIXTEENTH, ONE_EIGHTH, THREE_SIXTEENTHS, ONE_FOURTH,
+        FIVE_SIXTEENTHS, THREE_EIGHTHS, SEVEN_SIXTEENTHS, ONE_HALF, // end straw
+        THREE_FOURTHS, THREE_FOURTHS, THREE_FOURTHS, THREE_FOURTHS,
+        VoxelShapes.block(), VoxelShapes.block(), VoxelShapes.block(), VoxelShapes.block(), // end logs
+        VoxelShapes.block() // LIT
+    };
```

Comment:
```suggestion
    public static final VoxelShape[] SHAPES = Util.make(new VoxelShape[16], shapes -> {
        for (int i = 0; i < 8; i++)
        {
            shapes[i] = Block.box(0, 0, 0, 16, 1 + i, 16);
        }
        final VoxelShape half = Block.box(0, 0, 0, 16, 8, 16);
        final VoxelShape threeQuarters = Block.box(0, 0, 0, 16, 12, 16);
        for (int i = 0; i < 4; i++)
        {
            shapes[8 + i] = VoxelShapes.or(half, Block.box(0, 8, 0, 16, 12, 4));
            shapes[12 + i] = VoxelShapes.or(threeQuarters, Block.box(0, 12, 0, 16, 16, 4));
        }
    });
```

Double check the math + orientations of the "log" shapes but something like that should be cleaner and more accurate.